### PR TITLE
system check: improve arch

### PR DESCRIPTION
**Pull request by Chive**
Refactor system check architecture, improve error hinting (thanks to @kinkerl) and use ``busybox ping`` for docker server connection check

### CODE REVIEW

**Comment by stefanfoulis** [comment]
:+1:

**Comment by Chive** [comment]
Hey @kinkerl, could you please try running that branch on your machine and post results here?

**Comment by kinkerl** [comment]
The tests work when I am in the docker group. When I remove myself from the docker group and the error output looks good:
```
 ✓  Git
 ✓  Docker Client
 ✓  Docker Machine
 ✓  Docker Compose
 ✖  Docker Server Connectivity

The following errors occurred:
 Docker Server Connectivity:
 > Command 'docker run --rm busybox ping -c 1 8.8.8.8' returned non-zero exit status 1
 > No read permissions on /var/run/docker.sock. Please make sure the unix socket can be accessed without root permissions. More information can be found in the docker installation guide: https://docs.docker.com/engine/installation/linux/ubuntulinux/#create-a-docker-group

```

New structure is very nice. Two very minor things:
1) the error messages could be also printed with `fg='red'`
2) doc strings are missing (maybe not needed?)

One medium thing:
We could make better use of errorcodes. Right now, we only get the information 0 = no error, and 1= error. The check classes could get custom errorcodes (with default = 1) and the cli command could return these errorcode. This would help a lot to distinguish problems on shell level. This might be useful in the future when integrating aldryn-client into automated infrastructure. If multiple errors occur, the errorcode of the first found error should be returned.

**Comment by kinkerl** [comment]
One more thing. When reading the errormessage, it is not clear why `docker run --rm busybox ping -c 1 8.8.8.8`  was called in the first place and why it is a problem when its not working. It should be explained that this is a test to try connectivity to docker. 

In addition, this test checks for docker connectivity AND internet connectivity from inside the docker container. Is this intentional? Maybe this can be split into two tests:

1) `docker run --rm busybox true` <--- server connectivity
2) `docker run --rm busybox ping -c 1 8.8.8.8` <--- internet connectivity from inside docker

**Comment by kinkerl** [comment]
I had time to test the command in another network. 

My docker network was not set up properly due to a problem related to ipv4 and ipv6. This was found and tested with the ping command. But as already said, the server connection itself is working fine. Only the internet connection inside docker did not work. I created another test for this.

Then, later in the project setup, the npm install command inside docker failed because the DNS resolution needs a predefined DNS server in my network environment to work. So this was not tested properly and I also created a test for this:

Here is a commit with the changes. Is there a better way to share this in the pull request?
https://github.com/kinkerl/aldryn-client/commit/aaa993fd1b9cc6763ff1cb334481a7092288a6f7

Thoughts?

**Comment by Chive** [comment]
@kinkerl I've added some more changes, one for the new Docker for Mac Beta and I did some refactoring on the checks itself - could you please try to run them again?

You can either run it from source or use one of those binaries:
* [aldryn-Darwin-2.2a4.zip](https://github.com/aldryn/aldryn-client/files/208079/aldryn-Darwin-2.2a4.zip)
* [aldryn-Linux-2.2a4.zip](https://github.com/aldryn/aldryn-client/files/208084/aldryn-Linux-2.2a4.zip)

**Comment by Chive** [comment]
Another release with reduced timeouts.

* [aldryn-Darwin-2.2a5.zip](https://github.com/aldryn/aldryn-client/files/208290/aldryn-Darwin-2.2a5.zip)
* [aldryn-Linux-2.2a5.zip](https://github.com/aldryn/aldryn-client/files/208298/aldryn-Linux-2.2a5.zip)

**Comment by Chive** [comment]
Bugfix:

* [aldryn-Darwin-2.2a6.zip](https://github.com/aldryn/aldryn-client/files/208352/aldryn-Darwin-2.2a6.zip)

**Comment by kinkerl** [comment]
I don't have time to test it right now but you can try to use the timeout command (http://man7.org/linux/man-pages/man1/timeout.1.html). Not sure if this is part of busybox, it is in ubuntu. 

This will return a non-zero error code if the command had to be stopped after the timeout. If the command exits naturally it will return the error code of the command.

**Comment by Chive** [comment]
Thanks @kinkerl, that works :smile:

Yet another version:
* [aldryn-Linux-2.2a7.zip](https://github.com/aldryn/aldryn-client/files/208391/aldryn-Linux-2.2a7.zip)
* [aldryn-Darwin-2.2a7.zip](https://github.com/aldryn/aldryn-client/files/208393/aldryn-Darwin-2.2a7.zip)

**Comment by kinkerl** [comment]
perfect ;)

File: aldryn_client/utils.py [hunk 1]
```diff
@@ -155,6 +155,10 @@ def is_windows():
     return sys.platform == 'win32'
 
 
+def is_linux():
+    return sys.platform in ('linux', 'linux2')
```
Comment (GaretJax) [inline aldryn_client/utils.py]: Are these two values enough?

(here is how twisted does it https://github.com/twisted/twisted/blob/f074ba3d5083aa1503abcf194aece327e7f84805/twisted/python/runtime.py#L136-L143)

**Comment by GaretJax** [comment]
LGTM

File: aldryn_client/utils.py (hunk 1 above)
Comment (Chive) [inline aldryn_client/utils.py]: According to python docs, yes. But also according to python docs, one should use ``startswith``. Oh well.

**Pull request merged**